util/reflect.go

```diff
@@ -367,61 +413,368 @@ type NodeInfo struct {
+				// In the case of a map/slice, the path is of the form
+				// "container/element".
+				if IsTypeSlice(sf.Type) || IsTypeMap(sf.Type) {
+					nn.PathFromParent = p[0:1]
+				}
+				errs = AppendErrs(errs, forEachFieldInternal(nn, in, out, iterFunction))
+			}
 		}
 
-	case IsValueSlice(ni.FieldValue):
-		for i := 0; i < ni.FieldValue.Len(); i++ {
-			nn := *ni
-			nn.FieldValue = ni.FieldValue.Index(i)
-			errs = AppendErrs(errs, forEachFieldInternal(&nn, in, out, iterFunction))
+	case IsTypeSlice(t):
+		schema := *(ni.Schema)
+		schema.ListAttr = nil
+		var pp []string
+		// Leaf-list elements share the parent schema.
+		if !schema.IsLeafList() {
+			pp = []string{schema.Name}
+		}
+		if IsNilOrInvalidValue(v) {
+			// Traverse the type tree only from this point.
+			nn := &NodeInfo{
+				Parent:         ni,
+				PathFromParent: pp,
+				Schema:         &schema,
+				FieldValue:     reflect.Zero(t.Elem()),
+			}
+			errs = AppendErrs(errs, forEachFieldInternal(nn, in, out, iterFunction))
+		} else {
+			for i := 0; i < ni.FieldValue.Len(); i++ {
+				nn := *ni
+				// The schema for each element is the list schema minus the list
+				// attrs.
+				nn.Schema = &schema
+				nn.Parent = ni
+				nn.PathFromParent = pp
+				nn.FieldValue = ni.FieldValue.Index(i)
+				errs = AppendErrs(errs, forEachFieldInternal(&nn, in, out, iterFunction))
+			}
 		}
 
-	case IsValueMap(ni.FieldValue):
-		for _, key := range ni.FieldValue.MapKeys() {
-			nn := *ni
-			nn.FieldValue = ni.FieldValue.MapIndex(key)
-			nn.FieldKey = key
-			nn.FieldKeys = ni.FieldValue.MapKeys()
-			errs = AppendErrs(errs, forEachFieldInternal(&nn, in, out, iterFunction))
+	case IsTypeMap(t):
+		schema := *(ni.Schema)
+		schema.ListAttr = nil
+		if IsNilOrInvalidValue(v) {
+			nn := &NodeInfo{
+				Parent:         ni,
+				PathFromParent: []string{schema.Name},
+				Schema:         &schema,
+				FieldValue:     reflect.Zero(t.Elem()),
+			}
+			errs = AppendErrs(errs, forEachFieldInternal(nn, in, out, iterFunction))
+		} else {
+			for _, key := range ni.FieldValue.MapKeys() {
+				nn := *ni
+				nn.Schema = &schema
+				nn.Parent = ni
+				nn.PathFromParent = []string{schema.Name}
+				nn.FieldValue = ni.FieldValue.MapIndex(key)
+				nn.FieldKey = key
+				nn.FieldKeys = ni.FieldValue.MapKeys()
+				errs = AppendErrs(errs, forEachFieldInternal(&nn, in, out, iterFunction))
+			}
 		}
 	}
 
-	return nil
+	return errs
+}
+
+// GetNodes returns the nodes in the data tree at the indicated path, relative
+// to the supplied root. If for a list node is missing, all values in the list
+// are returned. If the key is partial, all nodes matching the values present
+// in the key are returned.
+// If the root is the tree root, the path may be absolute.
+// GetNodes returns an error if the path is not found in the tree, or an element
+// along the path is nil.
+func GetNodes(schema *yang.Entry, root interface{}, path *gpb.Path) ([]interface{}, []*yang.Entry, error) {
```

Comment:
Another comment on this function signature -- are elements in the`[]interface{}` meant to correspond to those in `[]*yang.Entry`. What happens in the case that we have a path that specifies a path to a `union` of `leafref` (allowable in Y1.1, not Y1.0). In this case, I think I can get multiple `[]*yang.Entry` schemas, do these correspond 1:1 to the entries in `[]interface{}`?